### DISCUSSION TITLE
gem installation completes without errors but does not compile native extensions

### DISCUSSION THREAD

**Discussion by Watita** [sparklemotion/sqlite3-ruby#545] [Q&A]
Hi everyone,
I am using debian testing with Ruby installed via apt (Ruby 3.1 and gem 3.4.20) and, since last update, the sqlite3 gem installs with no errors but the native extensions are not compiled and the require 'sqlite3' instruction fails with the following error:

irb(main):001:0> require 'sqlite3'
<internal:/usr/lib/ruby/vendor_ruby/rubygems/core_ext/kernel_require.rb>:86:in 'require': cannot load such file -- sqlite3/sqlite3_native (LoadError)
	from <internal:/usr/lib/ruby/vendor_ruby/rubygems/core_ext/kernel_require.rb>:86:in 'require'
	from /var/lib/gems/3.1.0/gems/sqlite3-2.0.2-x86_64-linux/lib/sqlite3.rb:6:in 'rescue in < top (required) >'
	from /var/lib/gems/3.1.0/gems/sqlite3-2.0.2-x86_64-linux/lib/sqlite3.rb:2:in `< top (required) >'
	from <internal:/usr/lib/ruby/vendor_ruby/rubygems/core_ext/kernel_require.rb>:160:in 'require'
	from <internal:/usr/lib/ruby/vendor_ruby/rubygems/core_ext/kernel_require.rb>:160:in 'rescue in require'
	from <internal:/usr/lib/ruby/vendor_ruby/rubygems/core_ext/kernel_require.rb>:40:in 'require'
	from (irb):1:in '< main >'
	from /usr/lib/ruby/gems/3.1.0/gems/irb-1.4.1/exe/irb:11:in '< top (required) >'
	from /usr/local/bin/irb:25:in 'load'
	from /usr/local/bin/irb:25:in '< main >'
<internal:/usr/lib/ruby/vendor_ruby/rubygems/core_ext/kernel_require.rb>:86:in 'require': /lib/x86_64-linux-gnu/libc.so: invalid ELF header - /var/lib/gems/3.1.0/gems/sqlite3-2.0.2-x86_64-linux/lib/sqlite3/3.1/sqlite3_native.so (LoadError)
	from <internal:/usr/lib/ruby/vendor_ruby/rubygems/core_ext/kernel_require.rb>:86:in 'require'
	from /var/lib/gems/3.1.0/gems/sqlite3-2.0.2-x86_64-linux/lib/sqlite3.rb:4:in '<top (required)>'
	from <internal:/usr/lib/ruby/vendor_ruby/rubygems/core_ext/kernel_require.rb>:160:in 'require'
	from <internal:/usr/lib/ruby/vendor_ruby/rubygems/core_ext/kernel_require.rb>:160:in 'rescue in require'
	from <internal:/usr/lib/ruby/vendor_ruby/rubygems/core_ext/kernel_require.rb>:40:in 'require'
	from (irb):1:in '< main >'
	from /usr/lib/ruby/gems/3.1.0/gems/irb-1.4.1/exe/irb:11:in '< top (required) >'
	from /usr/local/bin/irb:25:in 'load'
	from /usr/local/bin/irb:25:in '< main >'
<internal:/usr/lib/ruby/vendor_ruby/rubygems/core_ext/kernel_require.rb>:86:in 'require': cannot load such file -- sqlite3 (LoadError)
	from <internal:/usr/lib/ruby/vendor_ruby/rubygems/core_ext/kernel_require.rb>:86:in 'require'
	from (irb):1:in '< main >'
	from /usr/lib/ruby/gems/3.1.0/gems/irb-1.4.1/exe/irb:11:in '< top (required) >'
	from /usr/local/bin/irb:25:in 'load'
	from /usr/local/bin/irb:25:in '< main >'

I also attach here the output of 'gem install sqlite3 --verbose':

sergio@debian:~$ sudo gem install sqlite3 --verbose
HEAD https://index.rubygems.org/
200 OK
GET https://index.rubygems.org/info/sqlite3
200 OK
GET https://index.rubygems.org/quick/Marshal.4.8/sqlite3-2.0.2-x86_64-linux.gemspec.rz
403 Forbidden
/var/lib/gems/3.1.0/gems/sqlite3-2.0.2-x86_64-linux/.gemtest
/var/lib/gems/3.1.0/gems/sqlite3-2.0.2-x86_64-linux/CHANGELOG.md
/var/lib/gems/3.1.0/gems/sqlite3-2.0.2-x86_64-linux/CONTRIBUTING.md
/var/lib/gems/3.1.0/gems/sqlite3-2.0.2-x86_64-linux/FAQ.md
/var/lib/gems/3.1.0/gems/sqlite3-2.0.2-x86_64-linux/INSTALLATION.md
/var/lib/gems/3.1.0/gems/sqlite3-2.0.2-x86_64-linux/LICENSE
/var/lib/gems/3.1.0/gems/sqlite3-2.0.2-x86_64-linux/README.md
/var/lib/gems/3.1.0/gems/sqlite3-2.0.2-x86_64-linux/dependencies.yml
/var/lib/gems/3.1.0/gems/sqlite3-2.0.2-x86_64-linux/ext/sqlite3/aggregator.c
/var/lib/gems/3.1.0/gems/sqlite3-2.0.2-x86_64-linux/ext/sqlite3/aggregator.h
/var/lib/gems/3.1.0/gems/sqlite3-2.0.2-x86_64-linux/ext/sqlite3/backup.c
/var/lib/gems/3.1.0/gems/sqlite3-2.0.2-x86_64-linux/ext/sqlite3/backup.h
/var/lib/gems/3.1.0/gems/sqlite3-2.0.2-x86_64-linux/ext/sqlite3/database.c
/var/lib/gems/3.1.0/gems/sqlite3-2.0.2-x86_64-linux/ext/sqlite3/database.h
/var/lib/gems/3.1.0/gems/sqlite3-2.0.2-x86_64-linux/ext/sqlite3/exception.c
/var/lib/gems/3.1.0/gems/sqlite3-2.0.2-x86_64-linux/ext/sqlite3/exception.h
/var/lib/gems/3.1.0/gems/sqlite3-2.0.2-x86_64-linux/ext/sqlite3/extconf.rb
/var/lib/gems/3.1.0/gems/sqlite3-2.0.2-x86_64-linux/ext/sqlite3/sqlite3.c
/var/lib/gems/3.1.0/gems/sqlite3-2.0.2-x86_64-linux/ext/sqlite3/sqlite3_ruby.h
/var/lib/gems/3.1.0/gems/sqlite3-2.0.2-x86_64-linux/ext/sqlite3/statement.c
/var/lib/gems/3.1.0/gems/sqlite3-2.0.2-x86_64-linux/ext/sqlite3/statement.h
/var/lib/gems/3.1.0/gems/sqlite3-2.0.2-x86_64-linux/ext/sqlite3/timespec.h
/var/lib/gems/3.1.0/gems/sqlite3-2.0.2-x86_64-linux/lib/sqlite3.rb
/var/lib/gems/3.1.0/gems/sqlite3-2.0.2-x86_64-linux/lib/sqlite3/3.0/sqlite3_native.so
/var/lib/gems/3.1.0/gems/sqlite3-2.0.2-x86_64-linux/lib/sqlite3/3.1/sqlite3_native.so
/var/lib/gems/3.1.0/gems/sqlite3-2.0.2-x86_64-linux/lib/sqlite3/3.2/sqlite3_native.so
/var/lib/gems/3.1.0/gems/sqlite3-2.0.2-x86_64-linux/lib/sqlite3/3.3/sqlite3_native.so
/var/lib/gems/3.1.0/gems/sqlite3-2.0.2-x86_64-linux/lib/sqlite3/constants.rb
/var/lib/gems/3.1.0/gems/sqlite3-2.0.2-x86_64-linux/lib/sqlite3/database.rb
/var/lib/gems/3.1.0/gems/sqlite3-2.0.2-x86_64-linux/lib/sqlite3/errors.rb
/var/lib/gems/3.1.0/gems/sqlite3-2.0.2-x86_64-linux/lib/sqlite3/pragmas.rb
/var/lib/gems/3.1.0/gems/sqlite3-2.0.2-x86_64-linux/lib/sqlite3/resultset.rb
/var/lib/gems/3.1.0/gems/sqlite3-2.0.2-x86_64-linux/lib/sqlite3/statement.rb
/var/lib/gems/3.1.0/gems/sqlite3-2.0.2-x86_64-linux/lib/sqlite3/value.rb
/var/lib/gems/3.1.0/gems/sqlite3-2.0.2-x86_64-linux/lib/sqlite3/version.rb
Successfully installed sqlite3-2.0.2-x86_64-linux
1 gem installed

Thank you very much for your help

**Comment by flavorjones** [comment]
Hi, I'll try to help.

> the native extensions are not compiled

Well, you're installing the precompiled version, so you're not compiling, but the compiled extension is present. In your output, you can see:

> /var/lib/gems/3.1.0/gems/sqlite3-2.0.2-x86_64-linux/lib/sqlite3/3.0/sqlite3_native.so
> /var/lib/gems/3.1.0/gems/sqlite3-2.0.2-x86_64-linux/lib/sqlite3/3.1/sqlite3_native.so
> /var/lib/gems/3.1.0/gems/sqlite3-2.0.2-x86_64-linux/lib/sqlite3/3.2/sqlite3_native.so
> /var/lib/gems/3.1.0/gems/sqlite3-2.0.2-x86_64-linux/lib/sqlite3/3.3/sqlite3_native.so

However, this error message is buried in your output:

> internal:/usr/lib/ruby/vendor_ruby/rubygems/core_ext/kernel_require.rb:86:in 'require': /lib/x86_64-linux-gnu/libc.so: invalid ELF header - /var/lib/gems/3.1.0/gems/sqlite3-2.0.2-x86_64-linux/lib/sqlite3/3.1/sqlite3_native.so (LoadError)

This tells me that you're perhaps not on an x86_64-linux system, can you tell us more about your system? What version of rubygems are you running? Did you install this gem via bundler, or via `gem install`?

**Reply by flavorjones** [reply]
Or, you may want to compile the gem yourself, you can read how to do this in https://github.com/sparklemotion/sqlite3-ruby/blob/main/INSTALLATION.md if that's how you'd like to proceed.

**Reply by Watita** [reply]
Dear flavorjones,
thank you for your reply.
I have the same issue on two PCs, both of them with Debian testing x86_64 (see below the output of uname):

```
root@eon:~# uname -srm
Linux 6.9.7-amd64 x86_64
```
and this the version of rubygems:

```
root@eon:~# gem --version
3.4.20
```

I installed sqlite3 via `gem install sqlite3`

This issue raised after the latest `apt-get dist-upgrade` I did. Before this update everything was working fine, on both PCs.

Thank you again
Sergio

**Reply by flavorjones** [reply]
What is the output of `gem env` and/or `bundle env`?

**Reply by flavorjones** [reply]
There are a few variables that I'm trying to isolate:

- did something change in debian/testing
- did something change in the version of rubygems that you're running
- what was the previous state? did it use the "ruby" platform gem and compile it before, or was it using the precompiled gem before?
- does this reproduce if you build your own ruby (e.g. with ruby-build or ruby-install or rvm) and are on the latest rubygems?

**Reply by flavorjones** [reply]
When I run this in the docker:testing docker file, I have issues requiring the gem but I don't get the same error as you do, I get:

```
<internal:/usr/lib/ruby/vendor_ruby/rubygems/core_ext/kernel_require.rb>:86:in `require': libc.so: cannot open shared
 object file: No such file or directory - /var/lib/gems/3.1.0/gems/sqlite3-2.0.2-x86_64-linux/lib/sqlite3/3.1/sqlite3
_native.so (LoadError)
```

and I can't figure out what's going on, but it's a different problem than you.

**Reply by Watita** [reply]
This is the output of `gem env`:
```
RubyGems Environment:
  - RUBYGEMS VERSION: 3.4.20
  - RUBY VERSION: 3.1.2 (2022-04-12 patchlevel 20) [x86_64-linux-gnu]
  - INSTALLATION DIRECTORY: /var/lib/gems/3.1.0
  - USER INSTALLATION DIRECTORY: /home/sergio/.gem/ruby/3.1.0
  - RUBY EXECUTABLE: /usr/bin/ruby3.1
  - GIT EXECUTABLE: /usr/bin/git
  - EXECUTABLE DIRECTORY: /usr/local/bin
  - SPEC CACHE DIRECTORY: /home/sergio/.gem/specs
  - SYSTEM CONFIGURATION DIRECTORY: /etc
  - RUBYGEMS PLATFORMS:
     - ruby
     - x86_64-linux
  - GEM PATHS:
     - /var/lib/gems/3.1.0
     - /home/sergio/.gem/ruby/3.1.0
     - /usr/local/lib/ruby/gems/3.1.0
     - /usr/lib/ruby/gems/3.1.0
     - /usr/lib/x86_64-linux-gnu/ruby/gems/3.1.0
     - /usr/share/rubygems-integration/3.1.0
     - /usr/share/rubygems-integration/all
     - /usr/lib/x86_64-linux-gnu/rubygems-integration/3.1.0
  - GEM CONFIGURATION:
     - :update_sources => true
     - :verbose => true
     - :backtrace => true
     - :bulk_threshold => 1000
     - "gem" => "--no-document"
  - REMOTE SOURCES:
     - https://rubygems.org/
  - SHELL PATH:
     - /usr/local/bin
     - /usr/bin
     - /bin
     - /usr/local/games
     - /usr/games
```

I am pretty sure  that before the update the installation of sqlite3 was building the native extentions.

I also used rvm to install version 3.3.4 (the latest) and 3.3.1 again (that's the same I was using with the issue). But after the installation of 3.3.4 version, rubygems was updated to 3.5.9 and this time sqlite3 installs correctly without compiling native extentions, as you can see here:
```
sergio@debian:~$ rvm use 3.3.4
Using /home/sergio/.rvm/gems/ruby-3.3.4
sergio@debian:~$ gem install sqlite3
Successfully installed sqlite3-2.0.2-x86_64-linux-gnu
1 gem installed
sergio@debian:~$ irb
3.3.4 :001 > require 'sqlite3'
 => true 
```
and:
```
sergio@debian:~$ gem env
RubyGems Environment:
  - RUBYGEMS VERSION: 3.5.11
  - RUBY VERSION: 3.3.4 (2024-07-09 patchlevel 94) [x86_64-linux]
  - INSTALLATION DIRECTORY: /home/sergio/.rvm/gems/ruby-3.3.4
  - USER INSTALLATION DIRECTORY: /home/sergio/.gem/ruby/3.3.0
  - RUBY EXECUTABLE: /home/sergio/.rvm/rubies/ruby-3.3.4/bin/ruby
  - GIT EXECUTABLE: /usr/bin/git
  - EXECUTABLE DIRECTORY: /home/sergio/.rvm/gems/ruby-3.3.4/bin
  - SPEC CACHE DIRECTORY: /home/sergio/.gem/specs
  - SYSTEM CONFIGURATION DIRECTORY: /home/sergio/.rvm/rubies/ruby-3.3.4/etc
  - RUBYGEMS PLATFORMS:
     - ruby
     - x86_64-linux
  - GEM PATHS:
     - /home/sergio/.rvm/gems/ruby-3.3.4
     - /home/sergio/.rvm/rubies/ruby-3.3.4/lib/ruby/gems/3.3.0
  - GEM CONFIGURATION:
     - :update_sources => true
     - :verbose => true
     - :backtrace => true
     - :bulk_threshold => 1000
     - "gem" => "--no-document"
  - REMOTE SOURCES:
     - https://rubygems.org/
  - SHELL PATH:
     - /home/sergio/.rvm/gems/ruby-3.3.4/bin
     - /home/sergio/.rvm/gems/ruby-3.3.4@global/bin
     - /home/sergio/.rvm/rubies/ruby-3.3.4/bin
     - /home/sergio/.rvm/bin
     - /usr/local/bin
     - /usr/bin
     - /bin
     - /usr/local/games
     - /usr/games
```

So is it likely to be something wrong in the rubygems package that comes with apt-install /testing?

Thank you

**Reply by flavorjones** [reply]
> is it likely to be something wrong in the rubygems package that comes with apt-install /testing

This would be my guess, yes.
